README.md
# NEST

nest is a collection of tools. In particular, is a series of small command line tool.

now, it supported tools as:

- [x] `nest w`: Current location weather
- [x] `nest t`: Language translation
- [x] `nest b`: Bytes convert

### install
for install, just use pip:
```
pip install nestall
```
Then, execute on the command line:
```
nest --help
```

### tools
- `nest t`
    **nest t** provide translation service between diffrent languages, you just need to type `nest t <query>` at the command line, like this:


    ```
    > nest t hello
    ---------------------
    你好
    ---------------------
    ```

    type `nest t --help` for more feature！

- `nest w`
    provide weather report, just type `nest w <location>`：
    ```
    > nest w beijing
    beijing: ☀️ +9°C
    ```
- `nest b`
    provide betes and MB and GB covert to each other:
    ```
    > nest b 1024
    bytes: 1024, MB: 1.00, GB: 0.00
    > nest b -m 1024
    bytes: 1048576, MB: 1024.00, GB: 1.00
    > nest b -g 1
    bytes: 1048576, MB: 1024.00, GB: 1.00
    ```


HISTORY.md
## v0.1.1
update README
## v0.1.0
nest init


test/test_pdf.py
import os
from reportlab.platypus import SimpleDocTemplate, Paragraph, PageBreak
from reportlab.lib.styles import getSampleStyleSheet
from PyPDF2 import PdfFileReader
from nest.pdf import spilt_first_page, spilt_first_pages

here = os.path.abspath(os.path.dirname(__file__))


def create_pdf_file(content, target):
    my_pdf = SimpleDocTemplate(target)
    sample_style_sheet = getSampleStyleSheet()
    # add content
    flowables = []
    for c in content:
        flowables.append(Paragraph(c, style=sample_style_sheet["BodyText"]))
        flowables.append(PageBreak())

    my_pdf.build(flowables)
    target.close()


def test_create_pdf_file(tmpdir):
    content = ["page first", "page two", "page three", "page four", "page five"]
    target = tmpdir / 'target.pdf'
    with open(target, 'w+b') as file:
        create_pdf_file(content, file)

    with open(target, 'rb') as file:
        page_nums = PdfFileReader(file).getNumPages()

    assert page_nums == 5


def test_pdf_can_extract_first_page(tmpdir):
    content = ["page one", "page two", "page three", "page four", "page five"]
    fro = tmpdir / "fro.pdf"
    to = tmpdir / "to.pdf"

    with open(fro, 'w+b') as file:
        create_pdf_file(content, file)

    spilt_first_page(fro, to)

    with open(to, 'rb') as file:
        first_page = PdfFileReader(file).getPage(0).extractText()

    assert "page one" in first_page


def test_pdf_can_extract_first_pages(tmpdir):
    origin = tmpdir.mkdir("origin")
    target = tmpdir.mkdir("target")
    content = ["page first", "page two", "page three", "page four", "page five"]

    for i in range(1, 1001):
        with open(f'{origin}/{i}.pdf', 'w+b') as f:
            create_pdf_file(content, f)

    spilt_first_pages(origin, target)





test/test_weather.py
import pytest
from click.testing import CliRunner
from nest import wea_cli


class TestWeather:

    @pytest.fixture(autouse=True)
    def setup(self):
        self.runner = CliRunner()

    # def test_wea_cli_default(self):
    #     result = self.runner.invoke(wea_cli, ['beijing'])
    #     assert result.exit_code == 0
    #     assert 'beijing' in result.output
    

test/test_translate.py
import pytest
from click.testing import CliRunner
from nest import trans_cli


class TestTranslate:

    @pytest.fixture(autouse=True)
    def setup(self):
        self.runner = CliRunner()

    def test_trans_cli(self):
        result = self.runner.invoke(trans_cli, ['hello'])
        assert result.exit_code == 0
        assert '你好' in result.output
    
    def test_trans_cli_engine(self):
        result = self.runner.invoke(trans_cli, ['hello', '--engine=youdao'])
        assert 'youdao' in result.output

    def test_trans_cli_source_and_to(self):
        pass


nest/pdf.py
# """
# PDF is a series tools for pdf split、convert、merge and so on...
# """
# from functools import partial
# from time import sleep

# from PyPDF2 import PdfFileReader, PdfFileWriter
# from concurrent.futures import ThreadPoolExecutor
# import os
# import click
# from PyPDF2.utils import PdfReadError
# from tqdm import tqdm, trange
# from threading import RLock

# from tqdm.contrib.concurrent import thread_map


# def spilt_first_page(fro, to):
#     """ extract first page of pdf files """

#     with open(fro, 'rb') as f:
#         try:
#             pdf = PdfFileReader(f)
#         except PdfReadError as e:
#             return 'pdf read error'

#         # extract first pdf page
#         writer = PdfFileWriter()
#         try:
#             writer.addPage(pdf.getPage(0))
#         except PdfReadError as e:
#             return 'pdf read error'
#         # write to file
#         with open(to, "w+b") as to_f:
#             writer.write(to_f)

#     return "ok"


# def spilt_first_pages(origin, target):
#     pdf_files = [
#         (os.path.join(origin, name), os.path.join(target, name))
#         for name in os.listdir(origin)
#         if os.path.isfile(os.path.join(origin, name))
#     ]

#     thread_map(lambda x: spilt_first_page(*x), pdf_files)



# @click.command()
# @click.argument("origin")
# @click.argument("target")
# def pdf_split_cli(origin, target):
#     """this is a pdf tool."""
#     click.echo("\033[1;31m---------------------\033[0m")
#     spilt_first_pages(origin, target)


# if __name__ == "__main__":
#     pdf_split_cli()


nest/bytes.py
# byte is a tool convert bytes to MB or GB etc.

import typer
from typing import Optional

__version__ = "0.1.0"


def version_callback(value: bool):
    if value:
        typer.echo(f"Bytes {__version__}")
        raise typer.Exit()


def emphasize(s: str):
    return typer.style(s, fg=typer.colors.GREEN, bold=True)


def main(
    capacity: int,
    MB: bool = typer.Option(False, "--mb", "-m", help="convert MB to each other"),
    GB: bool = typer.Option(False, "--gb", "-g", help="convert GB to each other"),
    version: Optional[bool] = typer.Option(
        None, "--version", callback=version_callback, is_eager=True
    ),
):
    """
    Bytes is a tools to convert bytes and MB and GB each other.
    """
    if MB:
        b = capacity * 1024
        mb = capacity
        gb = capacity / 1024
    elif GB:
        b = capacity * 1024 * 1024
        mb = capacity * 1024
        gb = capacity
    else:
        b = capacity
        mb = capacity / 1024
        gb = capacity / 1024 / 1024

    bytes = emphasize("bytes: ")
    MB = emphasize("MB: ")
    GB = emphasize("GB: ")
    typer.echo(f"{bytes}{b}, {MB}{mb:.2f}, {GB}{gb:.2f}")


if __name__ == "__main__":
    typer.run(main)


nest/translate.py
import click
import requests
import uuid
import time
import hashlib

AUTH = {
    "youdao": {
        "APP_KEY": "218f4a5913a2b64d",
        "APP_SECRET": "tn49UOKE8VBy5zezLGWjr8HnusRMcxW8",
        "URL": "https://openapi.youdao.com/api",
    },
    "baidu": {
        "APP_KEY": "20200807000534875",
        "APP_SECRET": "4rpXNwhV0dpMaZFiXFui",
        "URL": "https://fanyi-api.baidu.com/api/trans/vip/translate",
    },
}


class BaseTranslate(object):
    def __init__(self, q, src, to):
        self.q = q
        self.src = src
        self.to = to

    @staticmethod
    def request(url: str, data: dict):
        headers = {"Content-Type": "application/x-www-form-urlencoded"}
        resp = requests.post(url=url, data=data, headers=headers)
        return resp.json()

    @staticmethod
    def status(s):
        return "\033[1m{0}\033[0m".format(s)


class YoudaoTranslate(BaseTranslate):
    def __init__(self, q, src, to):
        super().__init__(q, src, to)

        self.current_time = str(int(time.time()))
        self.salt = str(uuid.uuid4())
        self.sign = self.encrypt()

    def trans(self):

        trans_data = {
            "q": self.q,
            "from": self.src,
            "to": self.to,
            "appKey": AUTH["youdao"]["APP_KEY"],
            "salt": self.salt,
            "sign": self.sign,
            "signType": "v3",
            "curtime": self.current_time,
        }

        resp = self.request(AUTH["youdao"]["URL"], trans_data)

        result = []

        # base translation
        base_result = resp["translation"][0]

        # web translation and explains
        try:
            web_result = resp["web"]
            explains = resp["basic"]["explains"]
        except KeyError as e:
            web_result = ""
            explains = ""

        # if is word
        is_word = resp["isWord"]
        if is_word:
            result.append(base_result)
            result.append("\033[1;31m---\033[0m")
            result.append("\033[1;31m词性:\033[0m")
            for exp in explains:
                result.append(exp)
            result.append("\033[1;31m---\033[0m")
            result.append("\033[1;31m网络释义:\033[0m")
            for web in web_result:
                value = ",".join(web["value"])
                result.append(web["key"] + ":" + value)
        else:
            result.append(base_result)

        result = "\n".join(result)

        error_code = resp["errorCode"]

        if error_code != "0":
            if error_code == "102":
                result = "不支持的语言类型!"
            elif error_code == "103":
                result = "翻译文本过长!"
        return result

    def encrypt(self):
        sign_str = (
            AUTH["youdao"]["APP_KEY"]
            + self.truncate()
            + self.salt
            + self.current_time
            + AUTH["youdao"]["APP_SECRET"]
        )
        hash_algorithm = hashlib.sha256()
        hash_algorithm.update(sign_str.encode("utf-8"))
        return hash_algorithm.hexdigest()

    def truncate(self):
        if self.q is None:
            return None
        size = len(self.q)
        return (
            self.q
            if size <= 20
            else self.q[0:10] + str(size) + self.q[size - 10 : size]
        )


class BaiduTranslate(BaseTranslate):
    def __init__(self, q, src, to):
        super(BaiduTranslate, self).__init__(q, src, to)
        self.salt = str(uuid.uuid4())
        self.sign = self.encrypt()

    def trans(self):
        trans_data = {
            "q": self.q,
            "from": self.src,
            "to": self.to,
            "appid": AUTH["baidu"]["APP_KEY"],
            "salt": self.salt,
            "sign": self.sign,
        }

        resp = self.request(AUTH["baidu"]["URL"], trans_data)

        result = []

        dst = resp["trans_result"][0]["dst"]
        result.append(dst)

        return "\n".join(result)

    def encrypt(self):
        sign_str = (
            AUTH["baidu"]["APP_KEY"] + self.q + self.salt + AUTH["baidu"]["APP_SECRET"]
        )
        hash_algorithm = hashlib.md5()
        hash_algorithm.update(sign_str.encode("utf-8"))
        return hash_algorithm.hexdigest()


class GoogleTranslate(BaseTranslate):
    pass


# trans
def trans(q: str, engine="youdao", src="auto", to="auto") -> str:
    youdao = YoudaoTranslate(q, src, to)
    google = GoogleTranslate(q, src, to)
    baidu = BaiduTranslate(q, src, to)

    result = ""

    if engine == "youdao":
        result = youdao.trans()
    elif engine == "baidu":
        result = baidu.trans()
    elif engine == "google":
        result = google.trans()
    else:
        result = "仅支持有道&百度"

    return result


nest/main.py
# byte is a tool convert bytes to MB or GB etc.

import typer
from typing import Optional
from .translate import trans
import sys
import subprocess
import os

__version__ = "0.1.0"

app = typer.Typer()


@app.command(name="t")
def trans_cli(
    query: str,
    source: str = typer.Option("auto", "--source", "-s", help="source language"),
    target: str = typer.Option("auto", "--target", "-t", help="target launguage"),
    engine: str = typer.Option("baidu", "--engine", "-e", help="translate engine"),
):
    """command t is a translation tool."""
    typer.echo("\033[1;31m---------------------\033[0m")
    result = trans(query, src=source, to=target, engine=engine)
    typer.echo(result)
    typer.echo("\033[1;31m---------------------\033[0m")
    typer.echo("翻译结果来自: %s" % engine)


def version_callback(value: bool):
    if value:
        typer.echo(f"Bytes {__version__}")
        raise typer.Exit()


def emphasize(s: str):
    return typer.style(s, fg=typer.colors.GREEN, bold=True)


@app.command(name="b")
def bytes_cli(
    capacity: int,
    MB: bool = typer.Option(False, "--mb", "-m", help="convert MB to each other"),
    GB: bool = typer.Option(False, "--gb", "-g", help="convert GB to each other"),
    version: Optional[bool] = typer.Option(
        None, "--version", callback=version_callback, is_eager=True
    ),
):
    """
    command b is a bytes convert tool.
    """
    if MB:
        b = capacity * 1024
        mb = capacity
        gb = capacity / 1024
    elif GB:
        b = capacity * 1024 * 1024
        mb = capacity * 1024
        gb = capacity
    else:
        b = capacity
        mb = capacity / 1024
        gb = capacity / 1024 / 1024

    bytes = emphasize("bytes: ")
    MB = emphasize("MB: ")
    GB = emphasize("GB: ")
    typer.echo(f"{bytes}{b}, {MB}{mb:.2f}, {GB}{gb:.2f}")


@app.command("w")
def weather_cli(location: str):
    """
    command w is a weather query tool.
    """
    if location != "":
        location = "/" + location

    if sys.platform == "win32":
        # use powershell to exec
        s = subprocess.Popen(
            [
                "powershell.exe",
                'Invoke-RestMethod "wttr.in{}?lang=zh&format=3"'.format(location),
            ],
            stdout=sys.stdout,
        )
        s.communicate()
    else:
        os.system('curl "wttr.in{}?lang=zh&format=3"'.format(location))


if __name__ == "__main__":
    app()
